sparse-vectors-movies-reco/BLOG-DRAFT.md

```diff
@@ -0,0 +1,415 @@
+```
+
+Here, each rating is adjusted by subtracting the mean rating of the entire dataset and then dividing by the standard deviation. This technique effectively centers the data around 0 and scales it in terms of standard deviation. Such a transformation is particularly useful in sparse vector scenarios typical in recommendation systems, as it enhances model sensitivity to both positive and negative user preferences.
+
+```python
+# Normalize ratings
+
+# Sparse vectors can use advantage of negative values, so we can normalize ratings to have mean 0 and std 1
+# In this scenario we can take into account movies that we don't like
+
+ratings.rating = (ratings.rating - ratings.rating.mean()) / ratings.rating.std()
+```
+
+### Convert ratings to sparse vectors
+
+In this section, the code is focused on converting user movie ratings into sparse vectors. This is a necessary preprocessing step for many machine learning models, especially when dealing with large datasets as it helps in reducing memory usage and computation time. Sparse vectors only store non-zero elements, which is efficient given that not every user has rated every movie in a typical movie rating dataset.
+
+To accomplish this, the code employs a `defaultdict` from Python's collections module to create a dictionary (`user_sparse_vectors`) where each key represents a unique user ID. The values are dictionaries themselves, containing two lists: `values` for storing the ratings and `indices` for storing the corresponding movie IDs. This structure is chosen because a sparse vector can be effectively represented by just keeping track of the indices where there are non-zero elements (in this case, the movie IDs the user has rated) and the values at those indices (the user's ratings for those movies).
+
+The loop iterates over each row in the `ratings` data (presumably a DataFrame object holding user IDs, movie IDs, and ratings), appending the rating to the `values` list and the movie ID to the `indices` list for the corresponding user ID. This process effectively transforms the dense representation of the ratings into a sparse format, where the majority of the matrix, which would have been zeroes, is not explicitly stored.
+
+This transformation is particularly useful in scenarios where the dataset is vast and sparse, such as recommendation systems in which there are thousands of users and items but only a small fraction of all possible user-item pairs have interactions (ratings, views, purchases, etc.). Sparse representations help in significantly reducing the computational load and memory requirements for storing and processing this data.
+
+```python
+# Convert ratings to sparse vectors
+
+from collections import defaultdict
+
+user_sparse_vectors = defaultdict(lambda: {
+    "values": [],
+    "indices": []
+})
+
+for row in ratings.itertuples():
+    user_sparse_vectors[row.user_id]["values"].append(row.rating)
```

Comment:
compute `user_sparse_vectors[row.user_id]` only once and bind it to a variable :)